src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Generators/IVsSingleFileGeneratorFactoryAggregator.cs

```diff
@@ -0,0 +1,104 @@
+﻿using Microsoft.VisualStudio.ProjectSystem.VS.Utilities;
+using Microsoft.VisualStudio.Shell.Interop;
+using System;
+using System.Linq;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Generators
+{
+    internal abstract class IVsSingleFileGeneratorFactoryAggregator : OnceInitializedOnceDisposed, IVsSingleFileGeneratorFactory
+    {
+        // Constants for the generator information registry keys
+        private const string CLSIDKey = "CLSID";
+        private const string DesignTimeSourceKey = "GeneratesDesignTimeSource";
+        private const string SharedDesignTimeSourceKey = "GeneratesSharedDesignTimeSource";
+        private const string DesignTimeCompilationFlagKey = "UseDesignTimeCompilationFlag";
+
+        private readonly IServiceProvider _serviceProvider;
+        private readonly IVSRegistryHelper _registryHelper;
+        private IRegistryKey _settingsRoot;
+
+        public IVsSingleFileGeneratorFactoryAggregator(IServiceProvider serviceProvider, IVSRegistryHelper registryHelper)
+        {
+            Requires.NotNull(serviceProvider, nameof(serviceProvider));
+            Requires.NotNull(registryHelper, nameof(registryHelper));
+            _serviceProvider = serviceProvider;
+            _registryHelper = registryHelper;
+        }
+
+        public int CreateGeneratorInstance(string wszProgId, out int pbGeneratesDesignTimeSource, out int pbGeneratesSharedDesignTimeSource, out int pbUseTempPEFlag, out IVsSingleFileGenerator ppGenerate)
+        {
+            // The only user in the project system does not call this method.
+            throw new NotImplementedException();
+        }
+
+        public int GetDefaultGenerator(string wszFilename, out string pbstrGenProgID)
+        {
+            // The only user in the project system does not call this method.
+            throw new NotImplementedException();
+        }
+
+        public int GetGeneratorInformation(string wszProgId, out int pbGeneratesDesignTimeSource, out int pbGeneratesSharedDesignTimeSource, out int pbUseTempPEFlag, out Guid pguidGenerator)
+        {
+            Requires.NotNullOrEmpty(wszProgId, nameof(wszProgId));
+            EnsureInitialized(true);
+
+            pbGeneratesDesignTimeSource = 0;
+            pbGeneratesSharedDesignTimeSource = 0;
+            pbUseTempPEFlag = 0;
+            pguidGenerator = Guid.Empty;
+
+            var generatorKey = GetGeneratorKey(PackageGuid);
+            if (generatorKey == null)
+            {
+                return VSConstants.E_FAIL;
```

Comment:
The contract is to return E_FAIL? How do I know up-front if a generator exists for a given project?